File: Array/#228. Summary Ranges.cpp
class Solution {
public:
    vector<string> summaryRanges(vector<int>& nums) {
        int cnt=0, n = nums.size();
        if(nums.size()==0) return {};
        nums.push_back(nums[nums.size()-1]-2);
        vector<string> res; n++;
      
        for(int i=0; i<n-1; i++){
            if(nums[i]!=nums[i+1]-1){
                if(nums[i]==nums[i-cnt]) res.push_back(to_string(nums[i]));
                else res.push_back(to_string(nums[i-cnt])+"->"+to_string(nums[i]));
                cnt=0; continue;
            }
            cnt++;
        }
        
        return res;
    }
};

File: Array/#228. Summary Ranges.py
'''Method1
TC: O(n)
SC: O(1)
'''

class Solution:
    def summaryRanges(self, nums: List[int]) -> List[str]:
        n = len(nums)
        if n == 0:
            return []

        nums.append(nums[len(nums) - 1] + 2) # dummy data
        res = []
        start, prev = 0, 0 # index

        for i in range(1, len(nums)):
            if nums[i] != nums[prev] + 1:
                res.append(f"{nums[start]}->{nums[i - 1]}" if i - start > 1 else f"{nums[start]}")
                start = i
            prev = i

        return res